src/interactive-window/debugger/jupyter/kernelDebugAdapter.ts

```diff
@@ -82,14 +90,121 @@ export class KernelDebugAdapter extends KernelDebugAdapterBase {
+        }
+    }
+    protected override translateRealFileToDebuggerFile(
+        source: DebugProtocol.Source | undefined,
+        lines?: { line?: number; endLine?: number; lines?: number[] }
+    ) {
+        if (!source || !source.path || !lines || (typeof lines.line !== 'number' && !Array.isArray(lines.lines))) {
+            return;
+        }
+        const startLine = lines.line || lines.lines![0];
+        // Find the cell that matches this line in the IW file.
+        const cell = this.cellToDebugFileSortedInReverseOrderByLineNumber.find(
+            (item) => startLine >= item.metadata.interactive.lineIndex + 1
+        );
+        if (!cell) {
+            return;
+        }
+        source.path = cell.debugFilePath;
+        if (typeof lines?.endLine === 'number') {
+            lines.endLine = lines.endLine - (cell.lineOffset || 0);
+        }
+        if (typeof lines?.line === 'number') {
+            lines.line = lines.line - (cell.lineOffset || 0);
+        }
+        if (lines?.lines && Array.isArray(lines?.lines)) {
+            lines.lines = lines?.lines.map((line) => line - (cell.lineOffset || 0));
+        }
+    }
+
+    protected override async sendRequestToJupyterSession(message: DebugProtocol.ProtocolMessage) {
+        if (this.jupyterSession.disposed || this.jupyterSession.status === 'dead') {
+            traceInfo(`Skipping sending message ${message.type} because session is disposed`);
+            return;
+        }
+
+        const request = message as unknown as DebugProtocol.SetBreakpointsRequest;
+        if (request.type === 'request' && request.command === 'setBreakpoints') {
+            const sortedLines = (request.arguments.lines || []).concat(
+                (request.arguments.breakpoints || []).map((bp) => bp.line)
+            );
+            const startLine = sortedLines.length ? sortedLines[0] : undefined;
+            // Find the cell that matches this line in the IW file.
+            const cell = startLine
+                ? this.cellToDebugFileSortedInReverseOrderByLineNumber.find(
+                      (item) => startLine >= item.metadata.interactive.lineIndex + 1
+                  )
+                : undefined;
+            if (cell) {
+                const clonedRequest: typeof request = JSON.parse(JSON.stringify(request));
+                if (request.arguments.lines) {
+                    request.arguments.lines = request.arguments.lines.filter(
+                        (line) => line <= cell.metadata.generatedCode!.endLine
+                    );
+                }
+                if (request.arguments.breakpoints) {
+                    request.arguments.breakpoints = request.arguments.breakpoints.filter(
+                        (bp) => bp.line <= cell.metadata.generatedCode!.endLine
+                    );
+                }
+                if (sortedLines.filter((line) => line > cell.metadata.generatedCode!.endLine).length) {
```

Comment:
@rchiodo we add breakpoints for each cell by splitting the breakpoints by cells.

Reply:
Is breakpoints the only thing that needs this special translation? What about stack frames?

Reply:
that's fine